### PR TITLE
Add icelake as CPU architecture

### PR DESCRIPTION
**Pull request by masterleinad**
So far I didn't touch the `Makefile` because I don't quite understand what kind of variables I should set for icelake. In `CMake` we just try `-march=icelake-server -mtune=icelake-server` and don't set anything else.

### CODE REVIEW

**Review by dalg24 — COMMENTED** [review]
Please confirm that `KOKKOS_ENABLE_TM` does not need to be set

**Review by dalg24 — APPROVED** [review]

**Review by crtrott — CHANGES_REQUESTED** [review]
Why the change the AVX512? Note that there is unfortunately a difference between server grade SKX (which our setting uses) and non-server. The same for icelake. We probably need to distinguish.

File: Makefile.kokkos [hunk 1]
```diff
@@ -426,10 +427,9 @@ KOKKOS_INTERNAL_USE_ARCH_SSE42      := $(shell expr $(KOKKOS_INTERNAL_USE_ARCH_W
 KOKKOS_INTERNAL_USE_ARCH_AVX        := $(shell expr $(KOKKOS_INTERNAL_USE_ARCH_SNB) + $(KOKKOS_INTERNAL_USE_ARCH_AMDAVX))
 KOKKOS_INTERNAL_USE_ARCH_AVX2       := $(shell expr $(KOKKOS_INTERNAL_USE_ARCH_HSW) + $(KOKKOS_INTERNAL_USE_ARCH_BDW) + $(KOKKOS_INTERNAL_USE_ARCH_ZEN) + $(KOKKOS_INTERNAL_USE_ARCH_ZEN2) + $(KOKKOS_INTERNAL_USE_ARCH_ZEN3))
 KOKKOS_INTERNAL_USE_ARCH_AVX512MIC  := $(shell expr $(KOKKOS_INTERNAL_USE_ARCH_KNL))
-KOKKOS_INTERNAL_USE_ARCH_AVX512XEON := $(shell expr $(KOKKOS_INTERNAL_USE_ARCH_SKX))
```
[the remark this answers is not in the document]
Comment (dalg24) [reply Makefile.kokkos:429]: Would you prefer 
```
KOKKOS_INTERNAL_USE_ARCH_AVX512XEON := $(shell expr $(KOKKOS_INTERNAL_USE_ARCH_SKX) + $(KOKKOS_INTERNAL_USE_ARCH_ICX))
```
and then later
```
ifeq ($(KOKKOS_INTERNAL_USE_ARCH_AVX512MIC), 1)
  tmp := $(call kokkos_append_header,"$H""define KOKKOS_ARCH_AVX512XEON")
endif
```
?

[the remark this answers is not in the document]
Comment (crtrott) [reply Makefile.kokkos:429]: Wondering about making SKX the SKX512 or SKX_SERVER
 and ICX ICX512 or ICX_SERVER
The difference of the server variant is that it got AVX512
 the icelake-server gcc option probably reflects that. 
Then we could make SKX and ICX the none-server variant, which is I think how spack handles it?

[the remark this answers is not in the document]
Comment (masterleinad) [reply Makefile.kokkos:429]: In `Makefile.kokkos` `KOKKOS_INTERNAL_USE_ARCH_AVX512XEON` is only used to set flags for Skylake, though,
```
  KOKKOS_CXXFLAGS += -march=skylake-avx512 -mtune=skylake-avx512 -mrtm
```
Enabling it for Ice Lake wouldn't allow us to set the desired architecture flags. Hence, I decided to get rid of `KOKKOS_INTERNAL_USE_ARCH_AVX512XEON` altogether and just set the preprocessor variable `KOKKOS_ARCH_AVX512XEON` directly according to the architecture.

The alternative to this is what Damien is proposing.

**Comment by masterleinad** [comment]
> Why the change the AVX512? Note that there is unfortunately a difference between server grade SKX (which our setting uses) and non-server. The same for icelake. We probably need to distinguish.

The instruction set supported for Ice Lake is more similar to Skylake than to Knights Landing so it seemed appropriate to use `KOKKOS_ARCH_AVX512XEON` for Ice Lake as well (noting that it's not used anywhere in `Kokkos`).

The flags we are using for Skylake are the server ones (`march=skylake-avx512` vs. ` -march=skylake `) AFAICT so I thought that we also want to use the server ones for Ice Lake.

**Comment by masterleinad** [comment]
> Wondering about making SKX the SKX512 or SKX_SERVER
and ICX ICX512 or ICX_SERVER
The difference of the server variant is that it got AVX512
the icelake-server gcc option probably reflects that.
Then we could make SKX and ICX the none-server variant, which is I think how spack handles it?

https://spack.readthedocs.io/en/latest/basic_usage.html#support-for-specific-microarchitectures shows that `spack` has
```
GenuineIntel - x86_64
    nocona  nehalem   sandybridge  haswell    skylake  cannonlake      cascadelake
    core2   westmere  ivybridge    broadwell  mic_knl  skylake_avx512  icelake
```
as microarchitecture. Also, I don't know if we need to be concerned with backward compatibility. According to https://en.wikichip.org/wiki/intel/microarchitectures/ice_lake_(client) and https://en.wikichip.org/wiki/intel/microarchitectures/ice_lake_(server) both Ice Lake variants support AVX512.

**Comment by nmhamster** [comment]
Folks - SKX means Sky Lake Xeon (which is the server), SKL (Sky Lake) is the client part. It might be good to try to use this naming so we have similar terms to those that Intel uses. ICX and ICL follow similar patterns.

**Comment by crtrott** [comment]
OK I like Si's comment here.

**Comment by masterleinad** [comment]
> OK I like Si's comment here.

@crtrott I think in that case this pull request does the right thing (adding IceLake server as ICX to both the CMake build system and Makefile.kokkos; SKX continues to mean Skylake server).

**Review by crtrott — APPROVED** [review]

**Comment by crtrott** [comment]
@masterleinad can you resolve the conflict? We also should probably add SKL and ICL as Si mentioned but could do that in a different one.

**Comment by masterleinad** [comment]
> @masterleinad can you resolve the conflict? We also should probably add SKL and ICL as Si mentioned but could do that in a different one.

Done. I'll add `SKL` and `ICL` in a separate pull request.

**Comment by dalg24** [comment]
Retest this please

**Pull request merged**